src/sdk/pynni/nni/pbt_tuner/pbt_tuner.py

```diff
@@ -0,0 +1,342 @@
+                    x[key],
+                    is_rand,
+                    random_state,
+                    oldy[key] if oldy else None,
+                    Rand,
+                    name + "[%s]" % str(key)
+                )
+    elif isinstance(x, list):
+        y = list()
+        for i, x_i in enumerate(x):
+            if isinstance(x_i, dict):
+                if NodeType.NAME not in x_i.keys():
+                    raise RuntimeError('\'_name\' key is not found in this nested search space.')
+            y.append(json2parameter(
+                x_i,
+                is_rand,
+                random_state,
+                oldy[i] if oldy else None,
+                Rand,
+                name + "[%d]" % i
+            ))
+    else:
+        y = copy.deepcopy(x)
+    return y
+
+
+def exploit_and_explore(bot_trial_info, top_trial_info, factors, epoch):
+    """
+    Replace checkpoint of bot_trial with top, and perturb hyperparameters
+
+    """
+    bot_checkpoint_dir = bot_trial_info.checkpoint_dir
+    top_hyper_parameters = top_trial_info.hyper_parameters
+    hyper_parameters = copy.deepcopy(top_hyper_parameters)
+    for key in hyper_parameters.keys():
+        if key == 'load_checkpoint_dir':
+            hyper_parameters[key] = hyper_parameters['save_checkpoint_dir']
+        elif key == 'save_checkpoint_dir':
+            hyper_parameters[key] = os.path.join(bot_checkpoint_dir, str(epoch))
+        elif isinstance(hyper_parameters[key], float):
+            perturb = np.random.choice(factors)
+            hyper_parameters[key] *= perturb
+        else:
+            continue
+    bot_trial_info.hyper_parameters = hyper_parameters
+    bot_trial_info.clean_id()
+
+
+class Trial_Info:
+    """
+    Information of each trial, refresh for each epoch
+
+    """
+    def __init__(self, checkpoint_dir=None, hyper_parameters=None, parameter_id=None, score=None):
+        self.checkpoint_dir = checkpoint_dir
+        self.hyper_parameters = hyper_parameters
+        self.parameter_id = parameter_id
+        self.score = score
+
+    def clean_id(self):
+        self.parameter_id = None
+
+
+class PBTTuner(Tuner):
+    def __init__(self, optimize_mode="maximize", all_checkpoint_dir=None, population_size=10, factors=(1.2, 0.8), fraction=0.2):
+        """
+        Initialization
+
+        Parameters
+        ----------
+        optimize_mode: str
+            maximize or minimize
+        all_checkpoint_dir: str
+            directory to store training model checkpoint
+        population_size: int
+            number of trials for each epoch
+        factors: tuple
+            factors for perturbation
+        fraction: float
+            fraction for selecting bottom and top trials
+        """
+        self.optimize_mode = OptimizeMode(optimize_mode)
+        if all_checkpoint_dir is None:
+            experiment_id = os.getenv('NNI_EXP_ID')
+            all_checkpoint_dir = os.path.join("~/nni/checkpoint/", experiment_id)
```

Comment:
does it works when I using PAI?

Reply:
if the experiment is not local mode, users should provide a path in shared storage that can be accessed by all the trials.